src/manager/checkpoint.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use std::collections::hash_map::RandomState;
+use std::collections::{HashMap, HashSet};
+use std::time::Duration;
+
+use anyhow::Result;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use futures_util::future::join_all;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::MethodNum;
+use ipc_gateway::Checkpoint;
+use ipc_sdk::subnet_id::SubnetID;
+use primitives::TCid;
+use tokio::time::sleep;
+
+use crate::config::Subnet;
+use crate::lotus::client::LotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
+
+const CHAIN_HEAD_REQUEST_PERIOD: Duration = Duration::from_secs(10);
+
+/// Starts the checkpoint manager.
+#[allow(dead_code)]
+pub async fn start(subnets: HashMap<String, Subnet>) {
+    let managed_subnets = subnets_to_manage(subnets);
+
+    let mut futures = Vec::new();
+    for (child, parent) in managed_subnets {
+        futures.push(manage_subnet(child, parent));
+    }
+    join_all(futures).await;
+}
+
+/// This function takes a `HashMap<String, Subnet>` and returns a `Vec` of tuples of the form
+/// `(child_subnet, parent_subnet)`, where `child_subnet` is a subnet that we need to actively
+/// manage checkpoint for. This means that for each `child_subnet` there exists at least one account
+/// for which we need to submit checkpoints on behalf of to `parent_subnet`, which must also be
+/// present in the map.
+fn subnets_to_manage(subnets: HashMap<String, Subnet>) -> Vec<(Subnet, Subnet)> {
+    // First, we remap subnets by SubnetID.
+    let subnets_by_id: HashMap<SubnetID, Subnet> = subnets
```

Comment:
Just wondering, in other json rpc methods, converting to `HashMap<SubnetID, Subnet>` is also needed. Not sure if we can just update the config to `HashMap<SubnetID, Subnet>` instead of `Hashmap<String, Subnet>`

Reply:
This is a good point, but IIRC there was a reason why this couldn't be done. Let's wait for @hmoniz's input.

Reply:
In the config, each subnet is specified through a section declaration of the form `[subnet.C]`, where `C` is a human-readable name that we attribute to the subnet. The `C` is what ends up being the key for the `HashMap`. I, too, would prefer for the map to be of the form `<SubnetID, Subnet>` right off the bat. But for this to happen, we would have to instead declare each subnet section as, for instance, `[subnet./root/f0100]` and write some custom deserialization logic for it. I'm not sure if this is worth the loss in UX. Maybe it's better to simply have a convenience function that remaps from `<String, Subnet>` to `<SubnetID, Subnet`>. Wdyt?

Reply:
this works as well!

Reply:
Let's go with this approach of remapping the hashmap, and let's see if users get confused or not with the current config format and change accordingly.